jetty-server/src/main/java/org/eclipse/jetty/server/MultiPartFormInputStream.java

```diff
@@ -507,16 +526,23 @@ else if ("".equals(_config.getLocation()))
                 contentTypeBoundary = QuotedStringTokenizer.unquote(value(_contentType.substring(bstart, bend)).trim());
             }
 
-            parser = new MultiPartParser(handler, contentTypeBoundary);
+            parser = new MultiPartParser(new Handler(), contentTypeBoundary);
             byte[] data = new byte[_bufferSize];
             int len;
             long total = 0;
 
             while (true)
             {
+                synchronized (this)
+                {
+                    if (state != State.PARSING)
```

Comment:
Consider if we really need the PARSING state?

Reply:
I think we do because only between `PARSING` and `PARSED` do we need to go to the `ERROR` state before that we can go straight to `DELETED`.